src/least_squares.rs

```diff
@@ -516,85 +723,148 @@ pub fn solve_rolling_ols(
+                } else {
+                    // just add the new observations
+                    state.update(x_new, y_new, None, None);
+                }
+
+                // update new coefficients
+                coefficients_i = state.solve();
+                coefficients.slice_mut(s![i, ..]).assign(&coefficients_i);
+
+                // add new observations to deque
+                is_valid_history.push_back(i);
+
+                // check if window is saturated with enough valid observations
+                if !is_saturated {
+                    is_saturated = is_valid_history.len() == window_size;
+                }
             } else {
-                let x_update = x_new.insert_axis(Axis(0)).into_owned(); // 1 x K
-                xtx_inv = update_xtx_inv(&xtx_inv, &x_update, None);
-                xty = xty + &x_new * y[i];
+                // forward fill last coefficients
+                coefficients.slice_mut(s![i, ..]).assign(&coefficients_i);
             }
-            coefficients.slice_mut(s![i, ..]).assign(&xtx_inv.dot(&xty));
         }
     } else {
-        // update X.T X & X.T Y and solve normal equations at every time step
-        // assign warm-up coefficients
-        let coef_warmup = solve_normal_equations(&xtx, &xty, false);
-
-        coefficients
-            .slice_mut(s![min_periods - 1, ..])
-            .assign(&coef_warmup);
-
-        // Slide the window and update coefficients
+        // 2) handle 'skip' mode (this matches statsmodels: mode=drop)
         for i in min_periods..n {
             let i_start = i.saturating_sub(window_size);
-            // update XTX w/ latest data point
-            let x_new = x.row(i);
-
-            // Add new contributions
-            xtx += &outer_product(&x_new, &x_new);
-            xty = xty + &x_new * y[i];
-
-            // Subtract the previous contribution
-            if i > window_size - 1 {
+            let is_valid_i = is_valid[i];
+            let is_valid_i_start = is_valid[i_start];
+
+            let n_valid_window = is_valid[i_start + 1..i + 1]
+                .iter()
+                .filter(|&&valid| valid)
+                .count();
+
+            // do a rank 2 update
+            if is_valid_i {
+                let x_new = x.row(i);
+                let y_new = y[i];
+                // window is saturated and both ends are valid: do rolling [rank 2 update]
+                if (i >= window_size) & is_valid_i_start {
+                    let x_prev = x.row(i_start);
+                    let y_prev = y[i_start];
+                    state.update(x_new, y_new, Some(x_prev), Some(y_prev));
+                } else {
+                    // only new observation is valid : do expanding [rank 1 update]
+                    state.update(x_new, y_new, None, None);
+                }
+                // only update coefficients if the window contains at least min_period obs
+                if n_valid_window >= n_valid {
```

Comment:
@AdrienDart  - this now does what you described as "mixed" which I think makes a lot of sense. Effectively this does: 

for every fixed window, drop missing observations. If number of remaining valid observations >= min_periods: produce an estimate otherwise carry over (forward fill) last valid coefficients